testing.rs

```diff
@@ -0,0 +1,621 @@
+/// Inserts a suite of ink! unit tests intended for a contract implementing PSP22 trait.
+/// `$contract` argument should be the name of the contract struct.
+/// `$constructor` argument should be the name of a function, which initializes `$contract`
+/// with the given total supply of tokens.
+/// This macro should be invoked inside `#[ink::contract]` module.
+#[macro_export]
+macro_rules! tests {
+    ($contract:ident, $constructor:path) => {
+        use super::*;
+        use ink::env::{test::*, DefaultEnvironment as E};
+
+        type Event = <$contract as ::ink::reflect::ContractEventBase>::Type;
```

Comment:
just to avoid potential conflicts, you might want to put the whole code here into a new module, say `psp22_unit_tests`

Reply:
Totally right, done